File: dotnet/Allmark/Render/RenderUtils.cs
namespace Allmark.Render;

using Allmark.Types;

public static class RenderUtils
{
    public static void StartNewLine(MarkdownNode node, RendererState state)
    {
        if (state.Output.Length > 0 && node.Block)
        {
            if (state.Output.ToString().EndsWith('\n'))
            {
                state.Output.Length -= 1;
            }
            if (state.Output.ToString().EndsWith('\r'))
            {
                state.Output.Length -= 1;
            }
            state.Output.Append("\n");
        }
    }

    public static void InnerNewLine(MarkdownNode node, RendererState state)
    {
        if (node.Block && node.Children!.Count > 0 && node.Children?[0]?.Block == true)
        {
            state.Output.Append("\n");
        }
    }

    public static void EndNewLine(MarkdownNode node, RendererState state)
    {
        if (node.Block)
        {
            state.Output.Append("\n");
        }
    }
}

File: web/src/render/renderUtils.ts
import type MarkdownNode from "../types/MarkdownNode";
import type RendererState from "../types/RendererState";

export function startNewLine(node: MarkdownNode, state: RendererState): void {
	if (state.output.length && node.block) {
		if (state.output.endsWith("\n")) {
			state.output = state.output.slice(0, -1);
		}
		if (state.output.endsWith("\r")) {
			state.output = state.output.slice(0, -1);
		}
		state.output += "\n";
	}
}

export function innerNewLine(node: MarkdownNode, state: RendererState): void {
	if (node.block && node.children && node.children[0]?.block) {
		state.output += "\n";
	}
}

export function endNewLine(node: MarkdownNode, state: RendererState): void {
	if (node.block) {
		state.output += "\n";
	}
}
